lib/uri_parser.js

```diff
@@ -384,6 +402,74 @@ function parseQueryString(query, options) {
   return Object.keys(result).length ? result : null;
 }
 
+/**
+ * Checks a query string for invalid tls options according to the URI options spec.
+ *
+ * @param {string} queryString The query string to check
+ * @throws {MongoParseError}
+ */
+function checkTLSOptions(queryString) {
+  const queryStringKeys = Object.keys(queryString);
+  if (
+    queryStringKeys.indexOf('tlsInsecure') !== -1 &&
+    (queryStringKeys.indexOf('tlsAllowInvalidCertificates') !== -1 ||
+      queryStringKeys.indexOf('tlsAllowInvalidHostnames') !== -1)
+  ) {
+    throw new MongoParseError(
+      'The tlsInsecure option cannot be used with tlsAllowInvalidCertificates or tlsAllowInvalidHostnames.'
+    );
+  }
+
+  const queryStringHasTLS = queryStringKeys.indexOf('tls') !== -1;
+  const queryStringHasSSL = queryStringKeys.indexOf('ssl') !== -1;
+
+  if (queryStringHasTLS && queryStringHasSSL) {
+    let tlsValue;
+    if (Array.isArray(queryString['tls'])) {
+      tlsValue = queryString['tls'][0];
+    } else {
+      tlsValue = queryString['tls'];
+    }
+
+    let sslValue;
+    if (Array.isArray(queryString['ssl'])) {
+      sslValue = queryString['ssl'][0];
+    } else {
+      sslValue = queryString['ssl'];
+    }
+
+    if (tlsValue !== sslValue) {
+      throw new MongoParseError('All values of tls and ssl must be the same.');
+    }
+  }
+
+  if (queryStringHasTLS) {
+    checkAllOptions('tls', queryString);
+  }
+
+  if (queryStringHasSSL) {
+    checkAllOptions('ssl', queryString);
+  }
+}
+
+/**
+ * Checks a query string to ensure all tls/ssl options are the same.
+ *
+ * @param {string} key The key (tls or ssl) to check
+ * @param {string} queryString The query string to check
+ * @throws {MongoParseError}
+ */
+function checkAllOptions(key, queryString) {
```

Comment:
`checkAllOptions` doesn't seem to describe what's going on here. It seems like this can be rolled into the refactor proposed above, or otherwise maybe we can rename to like `assertTlsOptionsAreEqual` or something